modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/domain/DefaultDomainManager.java

```diff
@@ -21,65 +20,78 @@
-    if (domainsByDescriptor.containsKey(bundleDescriptorWrapper)) {
-      Domain domain = domainsByDescriptor.get(bundleDescriptorWrapper);
-      String availableVersion = domain.getDescriptor().getBundleDescriptor().getVersion();
-      String expectedVersion = descriptor.getVersion();
-      return isCompatibleVersion(availableVersion, expectedVersion);
+  public Domain getDomain(String domainName) throws DomainNotFoundException {
+    if (!domainsByName.containsKey(domainName)) {
+      throw new DomainNotFoundException(domainName, domainsByName.keySet());
     }
 
-    return false;
+    return domainsByName.get(domainName);
   }
 
   @Override
-  public void addDomain(Domain domain) {
-    final SemVerBundleDescriptorWrapper bundleDescriptorWrapper = new SemVerBundleDescriptorWrapper(domain.getDescriptor());
-    Domain foundDomain = domainsByDescriptor.get(bundleDescriptorWrapper);
-    if (foundDomain != null) {
-      throw new IllegalArgumentException(getDomainAlreadyExistsErrorMessage(getDomainName(domain), getDomainName(foundDomain)));
-    }
-    domainsByDescriptor.put(bundleDescriptorWrapper, domain);
+  public boolean contains(String name) {
+    return domainsByName.containsKey(name);
   }
 
   @Override
-  public void removeDomain(Domain domain) {
-    final SemVerBundleDescriptorWrapper bundleDescriptorWrapper = new SemVerBundleDescriptorWrapper(domain.getDescriptor());
-    domainsByDescriptor.remove(bundleDescriptorWrapper);
+  public Domain getCompatibleDomain(BundleDescriptor wantedDescriptor)
+      throws DomainNotFoundException, AmbiguousDomainReferenceException {
+    Domain foundDomain = null;
+
+    for (Domain domain : domainsByName.values()) {
+      BundleDescriptor currentDescriptor = domain.getDescriptor().getBundleDescriptor();
+      if (currentDescriptor == null) {
+        // Skip the domains without bundle descriptor (default, for example)
+        continue;
+      }
+
+      if (isCompatibleBundle(currentDescriptor, wantedDescriptor)) {
+        if (foundDomain != null) {
+          throw new AmbiguousDomainReferenceException(wantedDescriptor);
+        }
+        foundDomain = domain;
+      }
+    }
+
+    if (foundDomain == null) {
+      throw new DomainNotFoundException(wantedDescriptor.getArtifactFileName(), domainsByName.keySet());
+    }
+
+    return foundDomain;
   }
 
-  private static String getDomainName(Domain domain) {
-    DomainDescriptor domainDescriptor = domain.getDescriptor();
-    BundleDescriptor bundleDescriptor = domainDescriptor.getBundleDescriptor();
-    if (bundleDescriptor == null) {
-      return domainDescriptor.getName();
-    } else {
-      return bundleDescriptor.getArtifactFileName();
+  @Override
+  public boolean containsCompatible(BundleDescriptor descriptor) {
+    for (Domain domain : domainsByName.values()) {
+      if (isCompatibleBundle(domain.getDescriptor().getBundleDescriptor(), descriptor)) {
```

Comment:
I don't remember if the current application's logs are showing to which domain the application has been "linked" after it is deployed. Maybe we could show that if it is not present. Right?

Reply:
There is a "table" in the logs with application name, domain name, and status (deployed, failed...)